src/nbl/asset/ICPUDescriptorSet.cpp

```diff
@@ -0,0 +1,253 @@
+
+	return { infosBegin, infosBegin + count };
+}
+
+core::smart_refctd_ptr<IAsset> ICPUDescriptorSet::clone(uint32_t _depth) const
+{
+	auto layout = (_depth > 0u && m_layout) ? core::smart_refctd_ptr_static_cast<ICPUDescriptorSetLayout>(m_layout->clone(_depth - 1u)) : m_layout;
+	auto cp = core::make_smart_refctd_ptr<ICPUDescriptorSet>(std::move(layout));
+	clone_common(cp.get());
+
+	for (uint32_t t = 0u; t < static_cast<uint32_t>(IDescriptor::E_TYPE::ET_COUNT); ++t)
+	{
+		const auto type = static_cast<IDescriptor::E_TYPE>(t);
+
+		for (uint32_t i = 0u; i < m_descriptorInfos[t]->size(); ++i)
+		{
+			const auto& srcDescriptorInfo = m_descriptorInfos[t]->begin()[i];
+			auto& dstDescriptorInfo = cp->m_descriptorInfos[t]->begin()[i];
+
+			auto category = getCategoryFromType(type);
+			
+			if (category == IDescriptor::E_CATEGORY::EC_IMAGE)
+				dstDescriptorInfo.info.image = srcDescriptorInfo.info.image;
+			else
+				dstDescriptorInfo.info.buffer = srcDescriptorInfo.info.buffer;
+
+			if (_depth > 0u)
+			{
+				// Clone the descriptor.
+				{
+					assert(srcDescriptorInfo.desc);
+
+					IAsset* descriptor = nullptr;
+					if (category == IDescriptor::E_CATEGORY::EC_IMAGE)
+						descriptor = static_cast<ICPUImageView*>(srcDescriptorInfo.desc.get());
+					else if (category == IDescriptor::E_CATEGORY::EC_BUFFER_VIEW)
+						descriptor = static_cast<ICPUBufferView*>(srcDescriptorInfo.desc.get());
+					else
+						descriptor = static_cast<ICPUBuffer*>(srcDescriptorInfo.desc.get());
+
+					auto descriptorClone = descriptor->clone(_depth - 1);
+
+					if (category == IDescriptor::E_CATEGORY::EC_IMAGE)
+						dstDescriptorInfo.desc = core::smart_refctd_ptr_static_cast<ICPUImageView>(std::move(descriptorClone));
+					else if (category == IDescriptor::E_CATEGORY::EC_BUFFER_VIEW)
+						dstDescriptorInfo.desc = core::smart_refctd_ptr_static_cast<ICPUBufferView>(std::move(descriptorClone));
+					else
+						dstDescriptorInfo.desc = core::smart_refctd_ptr_static_cast<ICPUBuffer>(std::move(descriptorClone));
+
+				}
+
+				// Clone the sampler.
+				{
+					if ((category == IDescriptor::E_CATEGORY::EC_IMAGE) && srcDescriptorInfo.info.image.sampler)
+						dstDescriptorInfo.info.image.sampler = core::smart_refctd_ptr_static_cast<ICPUSampler>(srcDescriptorInfo.info.image.sampler->clone(_depth - 1u));
+				}
+			}
+			else
+			{
+				dstDescriptorInfo.desc = srcDescriptorInfo.desc;
+			}
+		}
+	}
+
+	return cp;
+}
+
+void ICPUDescriptorSet::convertToDummyObject(uint32_t referenceLevelsBelowToConvert)
+{
+	convertToDummyObject_common(referenceLevelsBelowToConvert);
+
+	if (referenceLevelsBelowToConvert)
+	{
+		--referenceLevelsBelowToConvert;
+		m_layout->convertToDummyObject(referenceLevelsBelowToConvert);
+
+		for (uint32_t t = 0u; t < static_cast<uint32_t>(IDescriptor::E_TYPE::ET_COUNT); ++t)
+		{
+			const auto type = static_cast<IDescriptor::E_TYPE>(t);
+			const auto descriptorCount = m_layout->getTotalDescriptorCount(type);
```

Comment:
you want to convert `m_layout` to dummy AFTER you need to use it for anything

it probably wont break but its good practice to convert to dummy in reverse order (like destructors work)

Reply:
Resolved.